Úkol 1
V souboru s daty máš data o hodnotách finančních indikátorů 100 největších společností obchodovaných na americké burze. Naším cílem je zjistit, které indikátory nejvíce ovlivňují cenu, a vytvořit model, který odhadne cenu akcie na základě hodnot finančních indikátorů.

In [37]:
import pandas as pd
import seaborn as sns
from scipy import stats
import statsmodels.formula.api as smf
from statsmodels.stats.outliers_influence import OLSInfluence
import numpy as np

data = pd.read_csv("df.csv")
data.head()

Company    CR    DA     FL     OPM      RT   TAT    DE     Y     PT  \
0   2Q3DU  1.19  0.20   2.21  0.0770    6.66  0.40  0.43  1.95   5.93   
1   46JNq  0.29  0.64  10.14  0.2037   20.04  0.36  6.52  7.26  43.48   
2   4ISOt  2.93  0.04   1.43  0.3055    6.55  0.72  0.06  7.78  18.38   
3   5AWwu  1.03  0.25   3.55  0.0708  146.43  1.80  0.88  6.17   5.15   
4   5hbZR  1.20  0.47   0.00  0.1677   30.92  0.93  0.00  0.00  17.06   

                   Sector  
0             Industrials  
1  Communication Services  
2  Communication Services  
3      Consumer Defensive  
4       Consumer Cyclical

In [38]:
data_corr = data.drop(columns=["Company", "Sector"])
data_corr.head()

CR    DA     FL     OPM      RT   TAT    DE     Y     PT
0  1.19  0.20   2.21  0.0770    6.66  0.40  0.43  1.95   5.93
1  0.29  0.64  10.14  0.2037   20.04  0.36  6.52  7.26  43.48
2  2.93  0.04   1.43  0.3055    6.55  0.72  0.06  7.78  18.38
3  1.03  0.25   3.55  0.0708  146.43  1.80  0.88  6.17   5.15
4  1.20  0.47   0.00  0.1677   30.92  0.93  0.00  0.00  17.06

Část 1
Vytvoř korelační matici a podívej se, který ukazatel má největší vliv na indikátor Y akcie.

In [39]:
data_corr.corr()

CR        DA        FL       OPM        RT       TAT        DE  \
CR   1.000000 -0.225917 -0.179232  0.284529 -0.151203 -0.123665 -0.141136   
DA  -0.225917  1.000000  0.340320  0.015032  0.206620 -0.083691  0.464786   
FL  -0.179232  0.340320  1.000000 -0.112430 -0.076335  0.156169  0.968612   
OPM  0.284529  0.015032 -0.112430  1.000000  0.051672 -0.298701 -0.050014   
RT  -0.151203  0.206620 -0.076335  0.051672  1.000000  0.206580 -0.065267   
TAT -0.123665 -0.083691  0.156169 -0.298701  0.206580  1.000000  0.094869   
DE  -0.141136  0.464786  0.968612 -0.050014 -0.065267  0.094869  1.000000   
Y   -0.018690  0.164599  0.759323  0.051308 -0.046290  0.301222  0.724604   
PT  -0.040483 -0.120993  0.032588  0.021409  0.097708  0.017872 -0.025348   

            Y        PT  
CR  -0.018690 -0.040483  
DA   0.164599 -0.120993  
FL   0.759323  0.032588  
OPM  0.051308  0.021409  
RT  -0.046290  0.097708  
TAT  0.301222  0.017872  
DE   0.724604 -0.025348  
Y    1.000000 -0.001363  
PT  -0.001363  1.000000

Největší vliv na Y má FL (Financial Leverage): 0.759323

Část 2
Vytvoř regresní model, který bude mít koeficient Y jako vysvětlovanou proměnnou. Do modelu vlož hodnoty ostatních indikátorů jako vysvětlující proměnné.
S využitím modulu statsmodels vytvoř regresní model a zobraz si tabulku se souhrnem významů. 

In [40]:
formula = "Y ~ CR + DA + FL + OPM + RT + TAT + DE + PT"
mod = smf.ols(formula=formula, data=data)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.664
Model:                            OLS   Adj. R-squared:                  0.635
Method:                 Least Squares   F-statistic:                     22.52
Date:                Mon, 13 Nov 2023   Prob (F-statistic):           1.45e-18
Time:                        18:08:06   Log-Likelihood:                -375.94
No. Observations:                 100   AIC:                             769.9
Df Residuals:                      91   BIC:                             793.3
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -11.0176      5.347     -2.061      0.042     -21.638      -0.397
CR             1.5014      1.240      1.211      0.229      -0.961       3.964
DA            -5.2206      9.370     -0.557      0.579     -23.834      13.392
FL             3.0104      1.060      2.840      0.006       0.905       5.116
OPM           23.4665      8.076      2.906      0.005       7.425      39.508
RT            -0.0063      0.021     -0.300      0.765      -0.048       0.035
TAT            8.4254      2.351      3.584      0.001       3.756      13.095
DE            -0.7283      2.066     -0.352      0.725      -4.833       3.376
PT            -0.0644      0.096     -0.669      0.505      -0.256       0.127
==============================================================================
Omnibus:                       29.136   Durbin-Watson:                   1.900
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              101.551
Skew:                           0.881   Prob(JB):                     8.88e-23
Kurtosis:                       7.611   Cond. No.                         592.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Podívej se na hodnoty koeficientů a na výsledky testu statistické významnosti koeficientů. Pokud je některý koeficient (nebo více koeficientů) nevýznamný, sestav nový model bez tohoto koeficientu (případně beze všech nevýznamných koeficientů).

--> CR, DA, RT, DE a PT jsou statisticky nevýznamné (p hodnota je větší než 0.05)

In [41]:
formula = "Y ~ FL + OPM + TAT"
mod = smf.ols(formula=formula, data=data)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.650
Model:                            OLS   Adj. R-squared:                  0.639
Method:                 Least Squares   F-statistic:                     59.51
Date:                Mon, 13 Nov 2023   Prob (F-statistic):           7.99e-22
Time:                        18:08:14   Log-Likelihood:                -377.99
No. Observations:                 100   AIC:                             764.0
Df Residuals:                      96   BIC:                             774.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -10.4188      2.865     -3.637      0.000     -16.106      -4.732
FL             2.5281      0.208     12.149      0.000       2.115       2.941
OPM           24.8478      7.542      3.295      0.001       9.878      39.818
TAT            8.5518      2.205      3.879      0.000       4.176      12.928
==============================================================================
Omnibus:                       26.748   Durbin-Watson:                   1.921
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               76.291
Skew:                           0.876   Prob(JB):                     2.71e-17
Kurtosis:                       6.904   Cond. No.                         50.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Pro všechna data odhadni ukazatel Y s využitím tvého modelu a odhadnuté ceny vlož do původní tabulky s daty. Dále vypočítej rozdíl mezi odhadem koeficientu a jeho skutečnou hodnotou.

In [42]:
data["Estimated Y"] = res.predict(data)
data["Variance"] = data["Y"] - data["Estimated Y"]
data.head()

Company    CR    DA     FL     OPM      RT   TAT    DE     Y     PT  \
0   2Q3DU  1.19  0.20   2.21  0.0770    6.66  0.40  0.43  1.95   5.93   
1   46JNq  0.29  0.64  10.14  0.2037   20.04  0.36  6.52  7.26  43.48   
2   4ISOt  2.93  0.04   1.43  0.3055    6.55  0.72  0.06  7.78  18.38   
3   5AWwu  1.03  0.25   3.55  0.0708  146.43  1.80  0.88  6.17   5.15   
4   5hbZR  1.20  0.47   0.00  0.1677   30.92  0.93  0.00  0.00  17.06   

                   Sector  Estimated Y   Variance  
0             Industrials     0.502359   1.447641  
1  Communication Services    23.356512 -16.096512  
2  Communication Services     6.944718   0.835282  
3      Consumer Defensive    15.708512  -9.538512  
4       Consumer Cyclical     1.701358  -1.701358

Najdi akcii, kde je tento rozdíl největší (tj. hledáme akcii, které náš model predikuje výrazně vyšší cenu, než jaká je ve skutečnosti, tato akcie je potenciálně na trhu podhodnocená).

In [43]:
min_var = data["Variance"].min()
data[data["Variance"] == min_var]

Company    CR    DA     FL     OPM     RT   TAT    DE     Y   PT  \
79   SYXLy  0.76  0.37  18.64  0.0631  12.44  0.41  6.93  6.97  3.4   

         Sector  Estimated Y   Variance  
79  Industrials    41.779545 -34.809545

Bonus 1: Sestav model s využitím robustní regrese. Opět proveď vyřazení koeficinetů, které nejsou statisticky významné, a sestav model pouze s významnými koeficienty. Vlož odhady cen do původních dat a opět najdi potenciálně nejvíce podhodnocenou akcii. Jde v případě robustní regese o stejnou akcii, nebo se akcie liší?

In [44]:
data_robust = pd.read_csv("df.csv")

In [45]:
formula = "Y ~ CR + DA + FL + OPM + RT + TAT + DE + PT"
mod = smf.rlm(formula=formula, data=data_robust)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                    Robust linear Model Regression Results                    
==============================================================================
Dep. Variable:                      Y   No. Observations:                  100
Model:                            RLM   Df Residuals:                       91
Method:                          IRLS   Df Model:                            8
Norm:                          HuberT                                         
Scale Est.:                       mad                                         
Cov Type:                          H1                                         
Date:                Mon, 13 Nov 2023                                         
Time:                        18:08:49                                         
No. Iterations:                    37                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -10.4961      3.453     -3.039      0.002     -17.265      -3.728
CR             2.0316      0.801      2.537      0.011       0.462       3.601
DA            -0.4762      6.052     -0.079      0.937     -12.338      11.386
FL             3.6694      0.685      5.360      0.000       2.328       5.011
OPM           16.3810      5.216      3.140      0.002       6.158      26.604
RT            -0.0029      0.014     -0.216      0.829      -0.029       0.024
TAT            5.2670      1.518      3.469      0.001       2.291       8.243
DE            -2.5507      1.335     -1.911      0.056      -5.166       0.065
PT            -0.0878      0.062     -1.412      0.158      -0.210       0.034
==============================================================================

If the model instance has been used for another fit with different fit parameters, then the fit options might not be the correct ones anymore .
"""

In [46]:
formula = "Y ~ CR + FL + OPM + TAT"
mod = smf.rlm(formula=formula, data=data_robust)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                    Robust linear Model Regression Results                    
==============================================================================
Dep. Variable:                      Y   No. Observations:                  100
Model:                            RLM   Df Residuals:                       95
Method:                          IRLS   Df Model:                            4
Norm:                          HuberT                                         
Scale Est.:                       mad                                         
Cov Type:                          H1                                         
Date:                Mon, 13 Nov 2023                                         
Time:                        18:09:03                                         
No. Iterations:                    23                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -9.9579      1.924     -5.176      0.000     -13.729      -6.187
CR             2.0977      0.681      3.081      0.002       0.763       3.432
FL             2.4422      0.123     19.840      0.000       2.201       2.683
OPM           13.9913      4.560      3.068      0.002       5.054      22.929
TAT            5.3858      1.290      4.176      0.000       2.858       7.913
==============================================================================

If the model instance has been used for another fit with different fit parameters, then the fit options might not be the correct ones anymore .
"""

In [47]:
data_robust["Estimated Y"] = res.predict(data_robust)
data_robust["Variance"] = data_robust["Y"] - data_robust["Estimated Y"]
min_var = data_robust["Variance"].min()
data_robust[data_robust["Variance"] == min_var]

Company    CR    DA     FL     OPM     RT   TAT    DE     Y   PT  \
79   SYXLy  0.76  0.37  18.64  0.0631  12.44  0.41  6.93  6.97  3.4   

         Sector  Estimated Y   Variance  
79  Industrials    40.250244 -33.280244

Je to stejná společnost.

Bonus 2: Použij původní (tedy "nerobustní") model a vyčísli Cookovu vzdálenost pro všechny hodnoty. Vyřaď všechny akcie s Cookovou vzdáleností vyšší než 1 a přepočítej regresní model. Nakonec opět najdi potenciálně nejvíce podhodnocenou akcii a podívej se, jestli jde o stejnou akcii jako u předchozích modelů.

In [48]:
data_cook = pd.read_csv("df.csv")

In [49]:
formula = "Y ~ FL + OPM + TAT"
mod = smf.ols(formula=formula, data=data_cook)
res = mod.fit()

influence = OLSInfluence(res)
data_cook["Cook Distance"] = influence.cooks_distance[0]
#data["Cook Distance"].max()


In [50]:
data_cook = data_cook[data_cook["Cook Distance"] < 1]

In [51]:
mod = smf.ols(formula=formula, data=data_cook)
res = mod.fit()

data_cook["Estimated Y"] = res.predict(data_cook)
data_cook["Variance"] = data_cook["Y"] - data_cook["Estimated Y"]
min_var = data_cook["Variance"].min()

In [52]:
data_cook[data_cook["Variance"] == min_var]

Company    CR    DA     FL     OPM     RT   TAT    DE     Y   PT  \
79   SYXLy  0.76  0.37  18.64  0.0631  12.44  0.41  6.93  6.97  3.4   

         Sector  Cook Distance  Estimated Y   Variance  
79  Industrials       0.336379    41.779545 -34.809545

Jde o stejnou společnost. V datech nebyla Cookova vzdálenost nikde větší než jedna.

Bonus 3: Přidej Sektor (poslední sloupec) jako vysvětlující proměnnou s využitím One Hot Encoding. Podívej se, které sektory mají kladný koeficient a které sektory mají záporný koeficient.

In [53]:
data_sector = pd.read_csv("df.csv")

In [54]:
data_sector["Sector"].value_counts()

Sector
Healthcare                20
Technology                20
Industrials               14
Consumer Cyclical         13
Communication Services    10
Consumer Defensive        10
Energy                     6
Basic Materials            3
Financial Services         3
Conglomerates              1
Name: count, dtype: int64

In [55]:
data_sector = pd.get_dummies(data=data_sector, prefix = "is", columns=["Sector"], dtype=int)
data_sector.head()

Company    CR    DA     FL     OPM      RT   TAT    DE     Y     PT  \
0   2Q3DU  1.19  0.20   2.21  0.0770    6.66  0.40  0.43  1.95   5.93   
1   46JNq  0.29  0.64  10.14  0.2037   20.04  0.36  6.52  7.26  43.48   
2   4ISOt  2.93  0.04   1.43  0.3055    6.55  0.72  0.06  7.78  18.38   
3   5AWwu  1.03  0.25   3.55  0.0708  146.43  1.80  0.88  6.17   5.15   
4   5hbZR  1.20  0.47   0.00  0.1677   30.92  0.93  0.00  0.00  17.06   

   is_Basic Materials  is_Communication Services  is_Conglomerates  \
0                   0                          0                 0   
1                   0                          1                 0   
2                   0                          1                 0   
3                   0                          0                 0   
4                   0                          0                 0   

   is_Consumer Cyclical  is_Consumer Defensive  is_Energy  \
0                     0                      0          0   
1                     0                      0          0   
2                     0                      0          0   
3                     0                      1          0   
4                     1                      0          0   

   is_Financial Services  is_Healthcare  is_Industrials  is_Technology  
0                      0              0               1              0  
1                      0              0               0              0  
2                      0              0               0              0  
3                      0              0               0              0  
4                      0              0               0              0

In [56]:
data_sector = data_sector.rename(columns={"is_Basic Materials":"is_Bas_Materials","is_Communication Services": "is_Com_Services", "is_Consumer Cyclical": "is_Con_Cyclical", "is_Consumer Defensive": "is_Con_Defensive","is_Financial Services": "is_Fin_Services"})
data_sector.columns

Index(['Company', 'CR', 'DA', 'FL', 'OPM', 'RT', 'TAT', 'DE', 'Y', 'PT',
       'is_Bas_Materials', 'is_Com_Services', 'is_Conglomerates',
       'is_Con_Cyclical', 'is_Con_Defensive', 'is_Energy', 'is_Fin_Services',
       'is_Healthcare', 'is_Industrials', 'is_Technology'],
      dtype='object')

In [57]:
formula = "Y ~ FL + OPM + TAT + is_Bas_Materials  + is_Com_Services + is_Conglomerates + is_Con_Cyclical + is_Con_Defensive + is_Energy + is_Fin_Services + is_Healthcare + is_Industrials + is_Technology"
mod = smf.ols(formula=formula, data=data_sector)
res = mod.fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.702
Model:                            OLS   Adj. R-squared:                  0.661
Method:                 Least Squares   F-statistic:                     17.08
Date:                Mon, 13 Nov 2023   Prob (F-statistic):           4.32e-18
Time:                        18:09:52   Log-Likelihood:                -369.98
No. Observations:                 100   AIC:                             766.0
Df Residuals:                      87   BIC:                             799.8
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept           -6.9599      2.967     -2.346      0.021     -12.857      -1.063
FL                   2.4726      0.208     11.873      0.000       2.059       2.887
OPM                 17.6464      8.265      2.135      0.036       1.219      34.073
TAT                  7.4755      2.424      3.084      0.003       2.658      12.293
is_Bas_Materials     0.2567      5.671      0.045      0.964     -11.015      11.528
is_Com_Services     -3.1696      3.372     -0.940      0.350      -9.873       3.533
is_Conglomerates    -3.8534      9.604     -0.401      0.689     -22.942      15.236
is_Con_Cyclical      3.9228      3.208      1.223      0.225      -2.454      10.299
is_Con_Defensive    -1.5535      3.698     -0.420      0.675      -8.904       5.797
is_Energy           -5.4862      4.169     -1.316      0.192     -13.772       2.799
is_Fin_Services      9.7416      6.040      1.613      0.110      -2.263      21.747
is_Healthcare       -2.7452      2.642     -1.039      0.302      -7.997       2.506
is_Industrials      -7.1937      2.991     -2.405      0.018     -13.139      -1.249
is_Technology        3.1205      2.626      1.188      0.238      -2.099       8.340
==============================================================================
Omnibus:                       30.481   Durbin-Watson:                   1.833
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              103.310
Skew:                           0.945   Prob(JB):                     3.69e-23
Kurtosis:                       7.607   Cond. No.                     1.13e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.7e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Sektory s kladným koeficientem: Basic Materials, Consumer Cyclical, Financial Services, Technology

Sektory se záporným koeficientem: Communication Services, Conglomerates, Consumer Defensive, Energy, Healthcare, Industrials

Statisticky významný se zdá být pouze Industrials. Některé kategorie jsou ale velmi málo zastoupené.